### ISSUE TITLE
IGNORE_WARNINGS does not work on AstropyDeprecationWarning? (sometimes)

### ISSUE THREAD

**Issue by pllim** [astropy/pytest-doctestplus#86]
In astropy/astropy#7928 , marking a line that emits `AstropyDeprecationWarning` with `+IGNORE_WARNINGS` caused failure in the `numpy-dev` job with `UnexpectedException` but it worked for the other ones. Example log: https://travis-ci.org/astropy/astropy/jobs/612691061

```
063 >>> bb = models.BlackBody1D()  # doctest: +IGNORE_WARNINGS
UNEXPECTED EXCEPTION: AstropyDeprecationWarning('BlackBody provides the same capabilities')
Traceback (most recent call last):
  File ".../python3.7/doctest.py", line 1329, in __run
    compileflags, 1), test.globs)
  File "<doctest astropy.modeling.blackbody.BlackBody1D[2]>", line 1, in <module>
  File ".../astropy/utils/decorators.py", line 122, in deprecated_func
    warnings.warn(message, category, stacklevel=2)
astropy.utils.exceptions.AstropyDeprecationWarning: BlackBody provides the same capabilities
.../astropy/modeling/blackbody.py:63: UnexpectedException
```

**Comment by ayshih** [comment]
Did you ever track down the root cause of your problem?  I think we're seeing something similar in SunPy, where a doctest is seemingly ignoring the `+IGNORE_WARNINGS` flag in just one of our CI builds, but the flag appears to work fine for that same doctest in all of the other builds.

**Comment by ayshih** [comment]
I just did a check, and `+SHOW_WARNINGS` is seemingly non-functional in that particular build as well.  It looks like the `warnings.catch_warnings` context manager for either directive is failing to catch warnings on this particular build.  My leading theory is that we're somehow tripping on `catch_warnings` not being thread-safe (see [note in docs](https://docs.python.org/3/library/warnings.html#warnings.catch_warnings)).  This particular build does use multiple workers (via `pytest-xdist`), but so do other builds of ours that don't have any issue...

**Comment by ayshih** [comment]
Better theory: this particular build is our remote-data build, so we may have threaded downloading that is causing threaded use of `catch_warnings`

**Comment by pllim** [comment]
Ah, thanks for tracking down the cause! Not sure what can be done about it though.

**Comment by ayshih** [comment]
It turns out that the root cause (at least for SunPy) is far more straightforward.  There's a bug in the handling of `IGNORE_WARNINGS`/`SHOW_WARNINGS` when `--remote-data=any` is specified.  See my fix in #152.

**Comment by saimn** [comment]
Fixed by #152
